Thanks for the quick response!

Regarding `getValue() == null`, you are absolutely right, `StringTag.getValue()` doesn't return null, that check was redundant on my end.

Regarding disconnecting on invalid strings: The client itself doesn't disconnect when interacting with the item; instead, the vanilla 1.21.2 client sends the `CONTAINER_CLICK` packet back to the server containing the `minecraft:lock` component with its `custom_name` string value set exactly to `"null"`. 

This happens consistently on our 1.12.2 server with 1.21.2 clients when players click on a Shulker Box inside a GUI (which likely has `Lock: ""` or `Lock: "null"` assigned by the GUI plugin to prevent opening). 

Because the vanilla client is the one sending this `"null"` string back to the server, `SerializerVersion.V1_20_5.toComponent()` parses `"null"` as `JsonNull`, which causes an uncaught `MergedCodecException` during `downgradeItemData` and drops the connection for legitimate players.

Here is the exact crash log we are getting when a player clicks the Shulker Box GUI item:

<details>
<summary>Click to expand Stacktrace</summary>

<pre><code>
[01:05:29 WARN]: [ViaVersion] ERROR IN Protocol1_21To1_21_2 IN REMAP OF CONTAINER_CLICK (0x10)
[01:05:29 WARN]: io.netty.handler.codec.DecoderException: com.viaversion.viaversion.exception.InformativeException: Please report the full error (!) from the console on the Via support Discord or open an issue on the relevant GitHub repository.
[01:05:29 WARN]: Source 0: com.viaversion.viaversion.protocols.v1_21to1_21_2.rewriter.BlockItemPacketRewriter1_21_2$$Lambda/0x00000000737a2e48
...
[01:05:29 WARN]: Caused by: com.viaversion.viaversion.libs.gson.JsonParseException: com.viaversion.viaversion.libs.mcstructs.converter.model.MergedCodecException: Failed to deserialize as either left or right: [Failed to deserialize as either left or right: [Expected string but got JsonNull],[Expected JsonArray but got JsonNull]],[Expected JsonObject but got JsonNull]
[01:05:29 WARN]:        at com.viaversion.viaversion.libs.mcstructs.converter.model.Result$Error.getOrThrow(Result.java:155)
[01:05:29 WARN]:        at com.viaversion.viaversion.libs.mcstructs.text.serializer.TextComponentCodec.deserialize(TextComponentCodec.java:243)
[01:05:29 WARN]:        at com.viaversion.viaversion.util.SerializerVersion.toComponent(SerializerVersion.java:94)
[01:05:29 WARN]:        at com.viaversion.viaversion.protocols.v1_21to1_21_2.rewriter.BlockItemPacketRewriter1_21_2.lambda$downgradeItemData$20(BlockItemPacketRewriter1_21_2.java:595)
[01:05:29 WARN]:        at com.viaversion.viaversion.api.minecraft.data.StructuredDataContainer.replace(StructuredDataContainer.java:175)
[01:05:29 WARN]:        at com.viaversion.viaversion.protocols.v1_21to1_21_2.rewriter.BlockItemPacketRewriter1_21_2.downgradeItemData(BlockItemPacketRewriter1_21_2.java:582)
[01:05:29 WARN]:        at com.viaversion.viaversion.protocols.v1_21to1_21_2.rewriter.BlockItemPacketRewriter1_21_2.handleItemToServer(BlockItemPacketRewriter1_21_2.java:485)
</code></pre>
</details>

I can update the PR to remove the redundant `getValue() == null` check and refine the `try-catch` if you agree that catching this `JsonParseException` is the right approach to prevent the server from kicking players for clicking GUI items!

--- a/common/src/main/java/com/viaversion/viaversion/protocols/v1_21to1_21_2/rewriter/BlockItemPacketRewriter1_21_2.java
+++ b/common/src/main/java/com/viaversion/viaversion/protocols/v1_21to1_21_2/rewriter/BlockItemPacketRewriter1_21_2.java
@@ -588,16 +588,11 @@ public final class BlockItemPacketRewriter1_21_2 extends StructuredItemRewriter<
 
             // Back from json in the string tag to plain text
             final StringTag customName = TagUtil.getNamespacedStringTag(itemComponentsTag, "custom_name");
-            if (customName == null || customName.getValue() == null || customName.getValue().equals("null")) {
+            if (customName == null || customName.getValue().isEmpty() || customName.getValue().equals("null")) {
                 return null;
             }
 
-            try {
-                return new StringTag(SerializerVersion.V1_20_5.toComponent(customName.getValue()).asUnformattedString());
-            } catch (Exception e) {
-                // If it fails to parse (e.g., malformed JSON), return empty lock to prevent client disconnect
-                return new StringTag("");
-            }
+            return new StringTag(SerializerVersion.V1_20_5.toComponent(customName.getValue()).asUnformattedString());
         });
         dataContainer.replace(StructuredDataKey.INSTRUMENT1_21_2, StructuredDataKey.INSTRUMENT1_20_5, instrument -> {
             if (instrument.hasId()) {

--- a/common/src/main/java/com/viaversion/viaversion/protocols/v1_21to1_21_2/rewriter/BlockItemPacketRewriter1_21_2.java
+++ b/common/src/main/java/com/viaversion/viaversion/protocols/v1_21to1_21_2/rewriter/BlockItemPacketRewriter1_21_2.java
@@ -588,7 +588,7 @@ public final class BlockItemPacketRewriter1_21_2 extends StructuredItemRewriter<
 
             // Back from json in the string tag to plain text
             final StringTag customName = TagUtil.getNamespacedStringTag(itemComponentsTag, "custom_name");
-            if (customName == null || customName.getValue().isEmpty() || customName.getValue().equals("null")) {
+            if (customName == null || customName.getValue().equals("null")) {
                 return null;
             }
 

